src/App.tsx

```diff
@@ -1,26 +1,241 @@
+  const [newTodo, setNewTodo] = useState('');
+  const [error, setError] = useState<string | null>(null);
+  const [filter, setFilter] = useState<'all' | 'active' | 'completed'>('all');
+  const [loading, setLoading] = useState<boolean>(false);
+  const [editingTodoId, setEditingTodoId] = useState<number | null>(null);
+  const [editingTodoTitle, setEditingTodoTitle] = useState<string>('');
+
+  useEffect(() => {
+    if (USER_ID) {
+      setLoading(true);
+      getTodos()
+        .then(setTodos)
+        .catch(() => setError('Unable to load todos'))
+        .finally(() => setLoading(false));
+    }
+  }, []);
+
+  const handleAddTodo = (event: React.FormEvent) => {
+    event.preventDefault();
+    if (!newTodo.trim()) {
+      setError('Title should not be empty');
+
+      return;
+    }
+
+    setLoading(true);
+
+    const newTempTodo: Todo = {
+      id: 0,
+      title: newTodo,
+      completed: false,
+      userId: USER_ID,
+    };
+
+    setTempTodo(newTempTodo);
+
+    createTodo(newTodo)
+      .then((newTodoItem: Todo) => {
+        setTodos(prevTodos => [...prevTodos, newTodoItem]);
+        setNewTodo('');
+        setTempTodo(null);
+      })
+      .catch(() => {
+        setError('Unable to add a todo');
+        setTempTodo(null);
+      })
+      .finally(() => setLoading(false));
+  };
+
+  const handleToggleTodo = (todo: Todo) => {
+    const updatedTodo = { ...todo, completed: !todo.completed };
+
+    setLoading(true);
+
+    updateTodo(todo.id, updatedTodo)
+      .then(() => {
+        setTodos(prevTodos =>
+          prevTodos.map(t => (t.id === todo.id ? updatedTodo : t)),
+        );
+      })
+      .catch(() => {
+        setError('Unable to update a todo');
+      })
+      .finally(() => setLoading(false));
+  };
+
+  const handleToggleAllTodos = () => {
+    const allCompleted = todos.every(todo => todo.completed);
+    const newStatus = !allCompleted;
+
+    setLoading(true);
+
+    const updatedTodos = todos.map(todo => ({
+      ...todo,
+      completed: newStatus,
+    }));
+
+    Promise.all(
+      updatedTodos.map(todo =>
+        updateTodo(todo.id, todo).catch(() => {
+          setError('Unable to update a todo');
+        }),
+      ),
+    )
+      .then(() => {
+        setTodos(updatedTodos);
+      })
+      .finally(() => setLoading(false));
+  };
+
+  const handleDeleteTodo = (todoId: number) => {
+    const deletedTodo = todos.find(todo => todo.id === todoId);
+
+    if (!deletedTodo) {
+      return;
+    }
+
+    setLoading(true);
+
+    deleteTodo(todoId)
+      .then(() => {
+        setTodos(prevTodos => prevTodos.filter(todo => todo.id !== todoId));
+      })
+      .catch(() => {
+        setError('Unable to delete a todo');
+      })
+      .finally(() => setLoading(false));
+  };
+
+  const handleEditTodo = (todo: Todo) => {
+    setEditingTodoId(todo.id);
+    setEditingTodoTitle(todo.title);
+  };
+
+  const handleUpdateTodo = (event: React.FormEvent) => {
```

Comment:
I'd pass `todoId` as an argument (you can get rid of the editingTodoId state here and just create isEditing state in TodoItem component) when calling it in the TodoItem component

also, you can create a wrapper function there (in TodoItem) to prevent default event